File: q2-backend/ts_template/roundupper_100.ts
import express from "express";

// location is the simple (x, y) coordinates of an entity within the system
// spaceCowboy models a cowboy in our super amazing system
// spaceAnimal models a single animal in our amazing system
type location = { x: number; y: number };
type spaceCowboy = { name: string; lassoLength: number };
type spaceAnimal = { type: "pig" | "cow" | "flying_burger" };

// spaceEntity models an entity in the super amazing (ROUND UPPER 100) system
type spaceEntity =
  | { type: "space_cowboy"; metadata: spaceCowboy; location: location }
  | { type: "space_animal"; metadata: spaceAnimal; location: location };

// === ADD YOUR CODE BELOW :D ===

const calcDistance = (location1: location, location2: location) => {
  return Math.sqrt(
    Math.pow(location1.x - location2.x, 2) +
      Math.pow(location1.y - location2.y, 2)
  );
};

// === ExpressJS setup + Server setup ===
const spaceDatabase = [] as spaceEntity[];
const app = express();
app.use(express.json());

// the POST /entity endpoint adds an entity to your global space database
app.post("/entity", (req, res) => {
  // TODO: fill me in
  let { entities } = req.body;
  for (const entity of entities) {
    spaceDatabase.push(entity);
  }
  res.status(200).send();
});

// lasooable returns all the space animals a space cowboy can lasso given their name
app.get("/lassoable", (req, res) => {
  let cowboy = spaceDatabase.filter(
    (entity: spaceEntity) =>
      entity.type === "space_cowboy" &&
      entity.metadata.name === req.body["cowboy_name"]
  )[0];

  let animals = spaceDatabase.filter(
    (entity: spaceEntity) =>
      entity.type === "space_animal" &&
      calcDistance(cowboy.location, entity.location) <=
        (cowboy.metadata as spaceCowboy).lassoLength
  );

  res.status(200).send({
    space_animals: animals.map((animal: spaceEntity) => ({
      type: (animal.metadata as spaceAnimal).type,
      location: animal.location,
    })),
  });
});

app.listen(8080);

File: q2-backend/py_template/roundupper_100.py
from dataclasses import dataclass
from enum import Enum
from typing import Union, NamedTuple, List
from flask import Flask, request
from math import sqrt, pow
# SpaceCowboy models a cowboy in our super amazing system
@dataclass
class SpaceCowboy:
    name: str
    lassoLength: int

# SpaceAnimal models a single animal in our amazing system
@dataclass
class SpaceAnimal:
    # SpaceAnimalType is an enum of all possible space animals we may encounter
    class SpaceAnimalType(Enum):
        PIG = "pig"
        COW = "cow"
        FLYING_BURGER = "flying_burger"

    type: SpaceAnimalType

# SpaceEntity models an entity in the super amazing (ROUND UPPER 100) system
@dataclass
class SpaceEntity:
    class Location(NamedTuple):
        x: int
        y: int

    metadata: Union[SpaceCowboy, SpaceAnimal]
    location: Location

# ==== HTTP Endpoint Stubs ====
app = Flask(__name__)
space_database: List[SpaceEntity] = []

# the POST /entity endpoint adds an entity to your global space database
@app.route('/entity', methods=['POST'])
def create_entity():
    entities = request.get_json()["entities"]
    for entity in entities:
        if (entity["type"] == 'space_cowboy'):
            space_database.append(SpaceEntity(SpaceCowboy(entity["metadata"]["name"], entity["metadata"]["lassoLength"]), SpaceEntity.Location(entity["location"]["x"], entity["location"]["y"])))
        else:
            space_database.append(SpaceEntity(SpaceAnimal(SpaceAnimal.SpaceAnimalType(entity["metadata"]["type"])), SpaceEntity.Location(entity["location"]["x"], entity["location"]["y"])))
    return {}, 200

# lasooable returns all the space animals a space cowboy can lasso given their name
@app.route('/lassoable', methods=['GET'])
def lassoable():
    cowboy = [cowboy for cowboy in space_database if type(cowboy.metadata) == SpaceCowboy and cowboy.metadata.name == request.get_json()["cowboy_name"]][0]
    animals = [animal for animal in space_database if type(animal.metadata) == SpaceAnimal and calcDistance(animal.location, cowboy.location) <= cowboy.metadata.lassoLength]
    return {"space_animals": [{"type": SpaceAnimal.SpaceAnimalType(animal.metadata.type).value, "location": {"x": animal.location.x, "y": animal.location.y} } for animal in animals]}, 200


def calcDistance(location1: SpaceEntity.Location, location2: SpaceEntity.Location) -> float:
    return sqrt(pow( location1.x - location2.x, 2) + pow(location1.y - location2.y, 2))


# DO NOT TOUCH ME, thanks :D
if __name__ == '__main__':
    app.run(debug=True, port=8080)